### PR TITLE
:tada: New Source:  WooCommerce

### PR DESCRIPTION
**Pull request by jacqueskpoty**
## What
As an Air-byte user, I needed to pull data from Woocommerce. Being that air-byte has no woocommerce source available now, I need to develop a source that supports data pulling from Woocommerce. #5122 

## How
I built a woocommerce source with the airbyte python CDK

## Recommended reading order
1. `x.java`
2. `y.python`

## Pre-merge Checklist
Expand the checklist which is relevant for this PR. 

<details><summary> <strong> Connector checklist </strong> </summary>
<p>

- [x] Issue acceptance criteria met
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [x] Secrets are annotated with `airbyte_secret` in the connector's spec
- [ ] Credentials added to Github CI if needed and not already present. [instructions for injecting secrets into CI](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#using-credentials-in-ci). 
- [x] Unit & integration tests added as appropriate (and are passing)
    * Community members: please provide proof of this succeeding locally e.g: screenshot or copy-paste acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] `/test connector=connectors/<name>` command as documented [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-an-existing-connector) is passing. 
    * Community members can skip this, Airbyters will run this for you. 
- [ ] Code reviews completed
- [x] Documentation updated 
    - [x] `README.md`
    - [x] `docs/SUMMARY.md` if it's a new connector
    - [x] Created or updated reference docs in `docs/integrations/<source or destination>/<name>`.
    - [ ] Changelog in the appropriate page in `docs/integrations/...`. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [x] `docs/integrations/README.md` contains a reference to the new connector
    - [ ] Build status added to [build page](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/builds.md)
- [ ] Build is successful
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] No major blockers
- [ ] PR merged into master branch
- [ ] Follow up tickets have been created
- [ ] Associated tickets have been closed & stakeholders notified
</p>
</details>

<details><summary> <strong> Connector Generator checklist </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Comment by jacqueskpoty** [comment]
Screen Shot 2021-07-31 at 3 11 29 AM

**Comment by marcosmarxm** [comment]
thanks @jacqueskpoty for the contribution

File: airbyte-integrations/connectors/source-woocommerce/build.gradle [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+dependencies {
+    implementation files(project(':airbyte-integrations:bases:source-acceptance-test').airbyteDocker.outputs)
+    implementation files(project(':airbyte-integrations:bases:base-python').airbyteDocker.outputs)
+}
+
+// used to allow local iteration to work
+task('installSingerTap', type: PythonTask) {
+    module = "pip"
+    command = "install woocommerce"
+}
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-woocommerce/build.gradle:20]: nit: end of line

File: airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    # Define primary key as sort key for full_refresh, or very first sync for incremental_refresh
+    primary_key = "id"
+    order_field = "date"
+
+    def __init__(self, shop: str, start_date: str, api_key: str, api_secret: str, **kwargs):
+            super().__init__(**kwargs)
+            self.start_date = start_date
+            self.shop = shop
+            self.api_key = api_key
+            self.api_secret = api_secret
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py:55]: 4 spaces

File: docs/integrations/sources/woocommerce.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+| `array` | `array` |  |
+| `object` | `object` |  |
+
+### Features
+
+| Feature | Supported?\(Yes/No\) | Notes |
+| :--- | :--- | :--- |
+| Full Refresh Sync | Yes |  |
+| Incremental - Append Sync | Yes |  |
+| Namespaces | No |  |
```
Comment (marcosmarxm) [inline docs/integrations/sources/woocommerce.md:43]: Maybe remove this part and add for each stream if it's possible to do incremental

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
A few comments about Python formatting and in docs.

File: airbyte-integrations/connectors/source-woocommerce/build.gradle (hunk 1 above)
Comment (jacqueskpoty) [reply airbyte-integrations/connectors/source-woocommerce/build.gradle:20]: Hi Marcos, I have to remove the `install woocommerce` , its not more needed. I see your comments however they are not explicit enough for me. Could you please add more details please.

File: airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py (hunk 2 above)
Comment (jacqueskpoty) [reply airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py:55]: Ok done

File: docs/integrations/sources/woocommerce.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    * Note: The UI will show all possible data sources and will show errors when syncing if it doesn't have permissions to access a resource.
+5. These two keys are your Consumer Key and Consumer Secret which respectively represent `Api key` and `Api secret` required for the integration
+6. You're ready to set up Woocommerce in Airbyte!
+
+
+## Changelog
+
+| Version | Date       | Pull Request | Subject |
+| :------ | :--------  | :-----       | :------ |
+| 0.1.0   | 2021-07-30 | [3787](https://github.com/airbytehq/airbyte/pull/3787) | Add woocommerce Source Connector |
```
[the remark this answers is not in the document]
Comment (jacqueskpoty) [reply docs/integrations/sources/woocommerce.md]: @bazarnov this is resolved

File: docs/integrations/sources/woocommerce.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+This source can sync data for the [Woocommerce API](http://woocommerce.github.io/woocommerce-rest-api-docs).
+
+This Source Connector is based on a [Airbyte CDK](https://docs.airbyte.io/contributing-to-airbyte/python).
+
+### Output schema
+
+This Source is capable of syncing the following core Streams:
+
+* [Abandoned Checkouts](https://help.shopify.com/en/api/reference/orders/abandoned_checkouts)
```
[the remark this answers is not in the document]
Comment (jacqueskpoty) [reply docs/integrations/sources/woocommerce.md]: @bazarnov This is resolved

File: docs/integrations/sources/woocommerce.md (hunk 3 above)
Comment (marcosmarxm) [reply docs/integrations/sources/woocommerce.md:35]: If possible yes.

File: airbyte-integrations/connectors/source-woocommerce/build.gradle (hunk 1 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-woocommerce/build.gradle:14]: Please add new line after last character.

File: airbyte-integrations/connectors/source-woocommerce/integration_tests/abnormal_state.json [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+  "customers": {
+    "date_modified": "2030-07-30T22:16:46"
+  },
+  "orders": {
+    "date_modified": "2050-07-30T22:16:46"
+  },
+  "coupons": {
+    "date_modified": "2060-07-30T22:16:46"
+  }
+}
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-woocommerce/integration_tests/abnormal_state.json:11]: add new line, please after last character, it's needed for correct formatting.

File: airbyte-integrations/connectors/source-woocommerce/source_woocommerce/schemas/TODO.md [hunk 7]
```diff
@@ -0,0 +1,25 @@
+# TODO: Define your stream schemas
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-woocommerce/source_woocommerce/schemas/TODO.md:1]: You can fairly remove this file, if you don't have dynamic json schema in your source-connector

**Comment by marcosmarxm** [comment]
@jacqueskpoty thanks again it's possible to run `./gradlew format` to apply formatting in all files?

@bazarnov is possible to review again?

**Comment by bazarnov** [comment]
Sure, will come back to it in a day.

File: airbyte-integrations/connectors/source-woocommerce/build.gradle [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+airbytePython {
+    moduleDirectory 'source_woocommerce'
+}
+
+dependencies {
+    implementation files(project(':airbyte-integrations:bases:source-acceptance-test').airbyteDocker.outputs)
+    implementation files(project(':airbyte-integrations:bases:base-python').airbyteDocker.outputs)
+}
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-woocommerce/build.gradle:14]: Need to `./gradlew format` or just add the new line after in the end of this file.

File: airbyte-integrations/connectors/source-woocommerce/integration_tests/sample_config.json [hunk 9]
```diff
@@ -0,0 +1,6 @@
+{
+  "shop": "pebblepost429763178.wpcomstaging",
+  "start_date": "2020-01-01",
+  "api_key": "ck_35ddd82891a6997e10ac7c3c5e2a0f4cdd958e30",
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-woocommerce/integration_tests/sample_config.json:4]: Are those credentials real? If so, they should not be here, please remove them ASAP.

File: airbyte-integrations/connectors/source-woocommerce/integration_tests/invalid_config.json [hunk 10]
```diff
@@ -0,0 +1,6 @@
+{
+  "shop": "SHOP_NAME",
+  "start_date": "START_DATE",
+  "api_key": "API_KEY",
+  "api_secret": "API_SECRET"
+}
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-woocommerce/integration_tests/invalid_config.json:6]: New Line

File: airbyte-integrations/connectors/source-woocommerce/source_woocommerce/schemas/orders.json [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+      }
+    },
+    "set_paid": {
+      "type": ["null", "boolean"]
+    },
+    "_links": {
+      "$ref": "_links.json"
+    }
+  }
+}
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-woocommerce/source_woocommerce/schemas/orders.json:279]: New line

File: airbyte-integrations/connectors/source-woocommerce/source_woocommerce/schemas/shared/_links.json [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+        "type": ["null", "object"],
+        "properties": {
+          "href": {
+            "type": ["null", "string"]
+          }
+        }
+      }
+    }
+  }
+}
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-woocommerce/source_woocommerce/schemas/shared/_links.json:27]: New line

File: airbyte-integrations/connectors/source-woocommerce/source_woocommerce/schemas/shared/billing.json [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+      "type": ["null", "string"]
+    },
+    "email": {
+      "type": ["null", "string"]
+    },
+    "phone": {
+      "type": ["null", "string"]
+    }
+  }
+}
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-woocommerce/source_woocommerce/schemas/shared/billing.json:38]: New line

File: airbyte-integrations/connectors/source-woocommerce/source_woocommerce/schemas/shared/meta_data.json [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+      },
+      "key": {
+        "type": ["null", "string"]
+      },
+      "value": {
+        "type": ["null", "string"]
+      }
+    }
+  }
+}
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-woocommerce/source_woocommerce/schemas/shared/meta_data.json:17]: new line

File: airbyte-integrations/connectors/source-woocommerce/source_woocommerce/schemas/shared/shipping.json [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+      "type": ["null", "string"]
+    },
+    "postcode": {
+      "type": ["null", "string"]
+    },
+    "country": {
+      "type": ["null", "string"]
+    }
+  }
+}
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-woocommerce/source_woocommerce/schemas/shared/shipping.json:32]: new line

File: airbyte-integrations/connectors/source-woocommerce/source_woocommerce/schemas/shared/taxes.json [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+      },
+      "shipping_tax_total": {
+        "type": ["null", "string"]
+      },
+      "meta_data": {
+        "$ref": "meta_data.json"
+      }
+    }
+  }
+}
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-woocommerce/source_woocommerce/schemas/shared/taxes.json:32]: new line

File: airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+        token = b64encode(b":".join((username, password))).strip().decode("ascii")
+        return token
+
+    def check_connection(self, logger, config) -> Tuple[bool, any]:
+        """
+        Testing connection availability for the connector.
+        """
+        shop = config["shop"]
+        api_version = "wc/v3"
+        api_key = config["api_key"]
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py:171]: No need to define `api_key` & `api_password` & `headers` here, if you call them directly from `config` later.

File: airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+        except requests.exceptions.RequestException as e:
+            return False, e
+
+    def streams(self, config: Mapping[str, Any]) -> List[Stream]:
+        """
+        TODO: Replace the streams below with your own streams.
+
+        :param config: A Mapping of the user input configuration as defined in the connector spec.
+        """
+        # TODO remove the authenticator if not required.
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py:191]: Remove this comment.

File: airbyte-integrations/connectors/source-woocommerce/source_woocommerce/spec.json [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+          "type": "string",
+          "description": "The name of the store. For https://EXAMPLE.com, the shop name is 'EXAMPLE'."
+        },
+      "start_date": {
+          "type": "string",
+          "description": "The date you would like to replicate data. Format: YYYY-MM-DD.",
+          "examples": ["2021-01-01"],
+          "pattern": "^[0-9]{4}-[0-9]{2}-[0-9]{2}$"
+        },
+      "api_key": {
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-woocommerce/source_woocommerce/spec.json:20]: This should have the property of `"airbyte_secret": true` as well.

File: airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+            self, stream_state: Mapping[str, Any] = None, stream_slice: Optional[Mapping[str, Any]] = None, **kwargs
+        ) -> Iterable[Mapping[str, Any]]:
+
+            slice = super().read_records(sync_mode = SyncMode.full_refresh, stream_slice = stream_slice,stream_state = stream_state)
+            yield from self.filter_records_newer_than_state(stream_state=stream_state, records_slice=slice)
+
+
+class Coupons(IncrementalWoocommerceStream):
+    data_field = "coupons"
+    order_field = "date"
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py:136]: You have it specified in base class, redefine `order_field` only if necessary. You can remove it from this class.

File: airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+    data_field = "customers"
+    order_field = "registered_date"
+
+    def path(self, **kwargs) -> str:
+        return f"{self.data_field}"
+
+
+class Orders(IncrementalWoocommerceStream):
+    data_field = "orders"
+    order_field = "date"
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py:151]: Can be removed, read comment above.

File: airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py [hunk 22]
```diff
@@ -0,0 +1,199 @@
...
+# Source
+class SourceWoocommerce(AbstractSource):
+
+    def _convert_auth_to_token(self, username: str, password: str) -> str:
+        username = username.encode("latin1")
+        password = password.encode("latin1")
+        token = b64encode(b":".join((username, password))).strip().decode("ascii")
+        return token
+
+    def check_connection(self, logger, config) -> Tuple[bool, any]:
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py:165]: Typehints:
`logger: AirbyteLogger, config: Mapping[str, Any]`

File: airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py [hunk 23]
```diff
@@ -0,0 +1,199 @@
...
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Tuple
+from urllib.parse import parse_qsl, urlparse
+
+import requests
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.sources.streams.http.auth import TokenAuthenticator
+
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py:37]: Add `from airbyte_cdk import AirbyteLogger`

File: airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py [hunk 24]
```diff
@@ -0,0 +1,199 @@
...
+            slice = super().read_records(sync_mode = SyncMode.full_refresh, stream_slice = stream_slice,stream_state = stream_state)
+            yield from self.filter_records_newer_than_state(stream_state=stream_state, records_slice=slice)
+
+
+class Coupons(IncrementalWoocommerceStream):
+    data_field = "coupons"
+    order_field = "date"
+
+    def path(self, **kwargs) -> str:
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py:138]: You can safely define the 
```
def path(self, **kwargs) -> str:
        return f"{self.data_field}"
```
Inside the `WoocommerceStream` class, and not to define it here. As far as you have the `data_field` defined per each stream.

File: airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py [hunk 25]
```diff
@@ -0,0 +1,199 @@
...
+        return f"{self.data_field}"
+
+class Customers(NonFilteredStream):
+    data_field = "customers"
+    order_field = "registered_date"
+
+    def path(self, **kwargs) -> str:
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py:145]: Same as above, about `def path(...)`

File: airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py [hunk 26]
```diff
@@ -0,0 +1,199 @@
...
+        return f"{self.data_field}"
+
+
+class Orders(IncrementalWoocommerceStream):
+    data_field = "orders"
+    order_field = "date"
+
+    def path(self, **kwargs) -> str:
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-woocommerce/source_woocommerce/source.py:153]: Same as above, about `def path(...)`

**Review by bazarnov — CHANGES_REQUESTED** [review]
Generally, very good! There are some work needs to be done, but it is rather small changes, than something else. Please consider to make these changes as soon as possible. 
Also, please remove or replace the sensitive information from the `/sample_config.json` file. It contains somebody's credentials)

File: airbyte-integrations/connectors/source-woocommerce/integration_tests/sample_config.json (hunk 9 above)
Comment (jacqueskpoty) [reply airbyte-integrations/connectors/source-woocommerce/integration_tests/sample_config.json:4]: They are just for test purposes. But Thanks

File: airbyte-integrations/connectors/source-woocommerce/build.gradle (hunk 8 above)
Comment (jacqueskpoty) [reply airbyte-integrations/connectors/source-woocommerce/build.gradle:14]: done

File: airbyte-integrations/connectors/source-woocommerce/integration_tests/sample_config.json (hunk 9 above)
Comment (jacqueskpoty) [reply airbyte-integrations/connectors/source-woocommerce/integration_tests/sample_config.json:4]: done

**Comment by jacqueskpoty** [comment]
@bazarnov Thanks for the review. All the changes requested have been taken care of. Let me know if there is something else to do.

**Review by bazarnov — APPROVED** [review]

**Comment by bazarnov** [comment]
@jacqueskpoty please, provide us with the successfully passed `acceptance-tests` on your side. Once they are, we will proceed with the future work on merging this great contribution.

**Comment by jacqueskpoty** [comment]
Screen Shot 2021-09-06 at 2 41 28 PM

File: airbyte-integrations/connectors/source-commercetools/source_commercetools/schemas/shared/taxed_item_price.json [hunk 27]
```diff
@@ -0,0 +1,11 @@
...
+  "type": ["null","object"],
+  "properties": {
+    "totalNet": {
+      "$ref": "money.json"
+    },
+    "totalGross": {
+      "$ref": "money.json"
+    }
+  }
+}
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-commercetools/source_commercetools/schemas/shared/taxed_item_price.json:11]: @jacqueskpoty you need to run `./gradlew format` to apply formatting. This is going to break master building. Could you provide this?

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-commercetools/source_commercetools/schemas/shared/taxed_item_price.json (hunk 27 above)
Comment (jacqueskpoty) [reply airbyte-integrations/connectors/source-commercetools/source_commercetools/schemas/shared/taxed_item_price.json:11]: @marcosmarxm This is normally part of commercetools source. Another source I contributed but has its own PR. So i will remove it from this PR so we can only deal with WooCommerce here

**Pull request closed without merging**